enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,302 @@
+default route on a different NIC so external traffic is segregated from
+internal traffic.
+
+TODO: I'd like to get a more specific user story since I know we have a
+number of requests for this.
+
+### Goals
+
+Ensure that all host networked services on a node have consistent interface
+and IP selection.
+
+### Non-Goals
+
+Support for platforms other than UPI (None).
+
+Complete support for multiple nics, with full control over what traffic
+gets routed where. However, this work should be able to serve as the basis
+for a broader multi-nic feature so we should avoid designs that would limit
+future work in this area.
+
+## Proposal
+
+The following is a possibly incomplete list of places that we do IP/interface
+selection. Ideally these should all use a single mechanism to do that so they
+are all consistent with each other.
+
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (local DNS) it may not matter if the IP selected is consistent
+with the other cases, but Node IP, configure-ovs, and Keepalived all need to
+match because their functionality depends on it. I'm less familiar with the
+requirements for Etcd, but it seems likely that should match as well. In
+general, it seems best if all IP selection logic comes from one place, whether
+it's strictly required or not.
+
+### Workflow Description
+
+At deployment time the cluster administrator will include a manifest that sets
+KUBELET_NODEIP_HINT appropriately. The nodeip-configuration service (which
+will now be set as a dependency for all other services that need IP/interface
+selection) will use that value to determine the desired IP and interface for
+all services on the node. It will write the results of this selection to a
+well-known location which the other services will consume. This way, we don't
+need to duplicate the selection logic to multiple places. It will happen once
+and be reused as necessary.
+
+#### Example
+
+- resolv-prepender has to run before any other node ip selection can take
+  place. Without resolv.conf populated the nodeip-configuration service cannot
+  pull the runtimecfg image.
+  - Note: This is not relevant for plain UPI deployments, but there are some
+    deployment mechanisms that use the IPI-style resolv.conf management. Since
+    this is not present in all deployments, we will likely need to leave this
+    as it is today. The prepender script uses the runtimecfg node-ip logic,
+    so although it won't be able to consume the output of nodeip-configuration
+    there won't be any duplicate logic either.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
```

Comment:
Saving files in /run will create issue with ovs-configuration as nodeip will find br-ex as interface and ovs will cleanup everything as a first step and only then will try to reconfigure. Maybe we should save those files in some other place and not to override interface file if br-ex was found? or we will need to split ovs logic and cleanup before nodeip?

Reply:
That's a good point, and now I'm really wondering if we should skip the interface part of this entirely and have it work solely on IPs. It's also a problem on first boot when the IP is still on the underlying interface. After configure-ovs runs the interface here will no longer be valid because the IP will have moved to br-ex.